File: src/lib/api/hosts.ts
import { invoke } from '@tauri-apps/api/core';

// ============================================================================
// Hosts Manager Types
// ============================================================================

/**
 * Status of the hosts file modification.
 * Matches Rust HostsStatus struct from hosts_manager.rs (camelCase)
 */
export interface HostsStatus {
    /** Whether Discord hosts entries are enabled */
    discordEnabled: boolean;
    /** Number of Discord host entries */
    discordEntriesCount: number;
    /** Whether a backup exists */
    backupExists: boolean;
    /** Whether hosts file is writable */
    isWritable: boolean;
    /** Path to the hosts file */
    hostsPath: string;
    /** Path to the backup file */
    backupPath: string;
    /** Last modified timestamp (ISO 8601) */
    lastModified: string | null;
    /** Backup timestamp (ISO 8601) */
    backupTimestamp: string | null;
}

// ============================================================================
// Hosts Manager API Functions
// ============================================================================

/**
 * Get the current hosts file status.
 * @returns Current hosts status including enabled state and entry count
 */
export async function getHostsStatus(): Promise<HostsStatus> {
    return invoke('get_hosts_status');
}

/**
 * Enable Discord hosts entries.
 * Adds IP mappings for Discord voice servers to bypass DPI blocking.
 * Creates a backup before modification.
 * Requires administrator privileges.
 * @throws Error if admin rights are required or hosts file is not writable
 */
export async function enableDiscordHosts(): Promise<void> {
    return invoke('enable_discord_hosts');
}

/**
 * Disable Discord hosts entries.
 * Removes Discord IP mappings from hosts file.
 * Requires administrator privileges.
 * @throws Error if admin rights are required or hosts file is not writable
 */
export async function disableDiscordHosts(): Promise<void> {
    return invoke('disable_discord_hosts');
}

/**
 * Toggle Discord hosts entries.
 * Enables if disabled, disables if enabled.
 * @param enable - Whether to enable or disable Discord hosts
 * @throws Error if admin rights are required
 */
export async function toggleDiscordHosts(enable: boolean): Promise<void> {
    if (enable) {
        return enableDiscordHosts();
    } else {
        return disableDiscordHosts();
    }
}

/**
 * Create a backup of the current hosts file.
 * Backup is stored in the app data directory.
 * @throws Error if backup creation fails
 */
export async function backupHosts(): Promise<void> {
    return invoke('backup_hosts');
}

/**
 * Restore hosts file from backup.
 * Requires administrator privileges.
 * @throws Error if backup doesn't exist or admin rights are required
 */
export async function restoreHosts(): Promise<void> {
    return invoke('restore_hosts');
}

/**
 * Flush DNS cache after hosts modification.
 * Runs `ipconfig /flushdns` to clear the Windows DNS resolver cache.
 * This ensures hosts file changes take effect immediately.
 * @throws Error if flush operation fails
 */
export async function flushDns(): Promise<void> {
    return invoke('flush_dns');
}

/**
 * Enable Discord hosts and flush DNS cache.
 * Convenience function that performs both operations.
 * @throws Error if any operation fails
 */
export async function enableDiscordHostsAndFlush(): Promise<void> {
    await enableDiscordHosts();
    await flushDns();
}

/**
 * Disable Discord hosts and flush DNS cache.
 * Convenience function that performs both operations.
 * @throws Error if any operation fails
 */
export async function disableDiscordHostsAndFlush(): Promise<void> {
    await disableDiscordHosts();
    await flushDns();
}

File: src-tauri/src/commands/hosts.rs
//! Hosts management commands
//!
//! Tauri IPC команды для управления системным hosts файлом.
//! Используется для добавления Discord voice серверов в hosts.

use std::process::Command;
use tracing::{info, warn};

use crate::core::errors::IsolateError;
use crate::core::hosts_manager::{self, HostsStatus};

/// Enable Discord hosts entries
///
/// Adds Discord voice server entries to the system hosts file.
/// Requires administrator privileges.
#[tauri::command]
pub async fn enable_discord_hosts() -> Result<(), IsolateError> {
    info!("Enabling Discord hosts entries");
    
    hosts_manager::add_discord_hosts().await
}

/// Disable Discord hosts entries
///
/// Removes Discord voice server entries from the system hosts file.
/// Requires administrator privileges.
#[tauri::command]
pub async fn disable_discord_hosts() -> Result<(), IsolateError> {
    info!("Disabling Discord hosts entries");
    
    hosts_manager::remove_discord_hosts().await
}

/// Get hosts status
///
/// Returns the current status of Discord hosts entries.
#[tauri::command]
pub async fn get_hosts_status() -> Result<HostsStatus, IsolateError> {
    info!("Getting hosts status");
    
    hosts_manager::get_hosts_status().await
}

/// Backup hosts file
///
/// Creates a backup of the current hosts file.
#[tauri::command]
pub async fn backup_hosts() -> Result<(), IsolateError> {
    info!("Creating hosts backup");
    
    hosts_manager::backup_hosts().await
}

/// Restore hosts from backup
///
/// Restores the hosts file from the backup.
/// Requires administrator privileges.
#[tauri::command]
pub async fn restore_hosts() -> Result<(), IsolateError> {
    info!("Restoring hosts from backup");
    
    hosts_manager::restore_hosts().await
}

/// Flush DNS cache
///
/// Clears the Windows DNS resolver cache using `ipconfig /flushdns`.
/// This should be called after modifying the hosts file to ensure
/// changes take effect immediately.
#[tauri::command]
pub async fn flush_dns() -> Result<(), IsolateError> {
    info!("Flushing DNS cache");
    
    // Run ipconfig /flushdns
    let output = Command::new("ipconfig")
        .arg("/flushdns")
        .output()
        .map_err(|e| IsolateError::Process(format!("Failed to run ipconfig: {}", e)))?;
    
    if output.status.success() {
        info!("DNS cache flushed successfully");
        Ok(())
    } else {
        let stderr = String::from_utf8_lossy(&output.stderr);
        warn!(stderr = %stderr, "Failed to flush DNS cache");
        Err(IsolateError::Process(format!(
            "ipconfig /flushdns failed: {}",
            stderr
        )))
    }
}
